### ISSUE TITLE
Web Debug Toolbar script is blocked by CSP

### ISSUE THREAD

**Issue by DavidPetrasek** [symfony/symfony#58223]
### Symfony version(s) affected

7.1.3

### Description

UPDATE: errors do occur even if CSP is is completely disabled (https://github.com/symfony/symfony/issues/58223#issuecomment-2360043257)

The CSP is handled by the NelmioSecurityBundle.
When navigating using UX Turbo Drive the WDT does not show up and these errors arise:
`Content-Security-Policy: The page’s settings blocked an inline script (script-src-elem) from being executed because it violates the following directive: “script-src https://127.0.0.1:5173/ 'nonce-Hie7bQmr1f4mayneR7+x+w==' 'unsafe-inline' 'nonce-1dca403eefc12cf2b625d960af5ce5c1'”`
Sometimes it also shows this error:
```
Uncaught TypeError: ajaxToolbarPanel is null
    renderAjaxRequests (index):521
    finishAjaxRequest (index):521
    load (index):521
    onreadystatechange (index):521
    request (index):521
    load (index):521
    loadToolbar (index):521
    finishAjaxRequest (index):521
    fetch (index):521
    promise callback*window.fetch (index):521
    fetchWithTurboHeaders turbo.es2017-esm.js:692
    perform turbo.es2017-esm.js:823
    setLater turbo.es2017-esm.js:1016
    setTimeout handler*setLater turbo.es2017-esm.js:1015
    LinkPrefetchObserver turbo.es2017-esm.js:3158
    LinkPrefetchObserver turbo.es2017-esm.js:3124
    start turbo.es2017-esm.js:3103
    start turbo.es2017-esm.js:5153
    start turbo.es2017-esm.js:5601
    <anonymous> turbo.es2017-esm.js:6578
```
![365524627-f86fb0af-3c71-4d02-89fa-4908dab1feab](https://github.com/user-attachments/assets/a61963c9-839e-49de-bbe8-b423a310d333)

And I have a related question: Why is the WDT generating a second nonce? Why not use the existing one?

### How to reproduce

symfony/web-profiler-bundle  7.1.3
nelmio/security-bundle             3.4.0 
symfony/ux-turbo                      2.18.0

DevSubscriber.php
```
<?php 
namespace App\EventSubscriber;

use Symfony\Component\DependencyInjection\Attribute\When;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\HttpKernel\Event\RequestEvent;
use Symfony\Component\HttpKernel\Event\ResponseEvent;
use Symfony\Component\HttpKernel\KernelEvents;


#[When(env: 'dev')]
class DevSubscriber implements EventSubscriberInterface
{
    private bool $isAjaxTurboDrive = false;
    
    public function onKernelRequest(RequestEvent $event)
    {
        if (!$event->isMainRequest()) {return;}
        
        $request = $event->getRequest();

        // Turbo
        $turboRequestID = $request->headers->get('x-turbo-request-id'); 
        if ($turboRequestID) 
        {
            $this->isAjaxTurboDrive = true;
        }
    }
    
    public function onKernelResponse(ResponseEvent $event)
    {                                                    
        $response = $event->getResponse();

        // These two lines make the WDT to show up, but errors remain. The nonce value is weak for testing purposes.
        $response->headers->set('X-SymfonyProfiler-Style-Nonce', 'fffff');    
        $response->headers->set('X-SymfonyProfiler-Script-Nonce', 'fffff');

        if ($this->isAjaxTurboDrive)
        {
            $response->headers->set('Symfony-Debug-Toolbar-Replace', '1');
        }
    }

    public static function getSubscribedEvents() : array
    {
        return [
            KernelEvents::REQUEST => 'onKernelRequest',
            KernelEvents::RESPONSE => 'onKernelResponse',
        ];
    }
}
?>
```

excerpt from nelmio_security.yaml:
```
...
csp:
        hosts: []
        content_types: []
        enforce:
            level1_fallback: false
            browser_adaptive:
                enabled: false
            # report-uri: '%router.request_context.base_url%/nelmio/csp/report'
            default-src:
                - 'none'
            style-src:
                - 'self'
                - https://www.gstatic.com https://fonts.googleapis.com                                # Google Fonts
            font-src:
                - 'self'
                - https://fonts.googleapis.com https://fonts.gstatic.com                              # Google Fonts
            img-src:
                - 'self' 
                - 'data:'
            form-action:
                - 'self'
            frame-src:
                - 'self'
                - 'blob:'
                - https://www.google.com
            frame-ancestors:
                - 'self'
            object-src:
                - 'self'
            media-src:
                - 'self'
            connect-src:
                - 'self' 
                - 'wss:'
            block-all-mixed-content: true # defaults to false, blocks HTTP content over HTTPS transport
            # upgrade-insecure-requests: true # defaults to false, upgrades HTTP requests to HTTPS transport
...
```

### Possible Solution

I managed to make the toolbar show up by setting both `X-SymfonyProfiler-Style-Nonce` and `X-SymfonyProfiler-Script-Nonce` (see DevSubscriber.php in "How to reproduce" section). However, I am still encountering the same errors in the console.

So possible solution is to **make WDT reuse the same nonce when UX Turbo is installed**. Here it is in more detail:
In method `getNonces` (in the ContentSecurityPolicyHandler) there it will check if the current request is coming from Turbo Drive:
`$request->headers->has('x-turbo-request-id');` 
a) If not, then it presumes this request is an initial page load and continues to check if UX Turbo is installed and if yes, it sets a new nonce and saves it to session.
b) If this is a Turbo Drive request, then it will reuse nonce from the session. **And here it should also trigger update of the WDT by setting the `Symfony-Debug-Toolbar-Replace` header (https://symfony.com/doc/current/profiler.html#updating-the-web-debug-toolbar-after-ajax-requests).**

**Comment by stof** [comment]
Using Turbo with Symfony does not require `Symfony-Debug-Toolbar-Replace` when morphing is disabled, because the script loading the toolbar is injected in the `<body>` and so is re-executed for each page load. This is part of the Turbo Drive behavior of being a drop-in progressive enhancement as documented in https://turbo.hotwired.dev/handbook/building#working-with-script-elements.
I haven't tried Turbo Morphing yet (I'm still using Turbo 7). If it breaks the documented behavior, this should be reported to the hotwired team IMO.

Regarding the issue with nonce and Turbo, this is a general issue with Turbo: by taking the content of a *new* page and inserting it in the existing page (as far as the browser is concerned), this content becomes subject to the CSP rules of the existing page instead of the setting sets in the response. Using CSP nonces with Turbo requires special logic both server-side and client-side to allow reusing previous nonces instead of generating new ones. See the [various issues on Turbo about CSP](https://github.com/hotwired/turbo/issues?q=is%3Aissue+CSP), while there is no official documentation about achieving it.
Anyway, I think the UX Turbo package should be the one setting the appropriate headers to make the integration work (as it will likely also be the one shipping the necessary JS code for the client-side)

**Comment by smnandre** [comment]
I'm not sure to be honest, as a standard Symfony + UX Turbo install works without any problem (web debug toolbar included) 

AFAIK, it's only happening with the aditional rules configured in the nelmio security bundle ...  🤷‍♂️

**Comment by DavidPetrasek** [comment]
@stof I am not using morphing

>  requires special logic both server-side and client-side to allow reusing previous nonces instead of generating new ones

I have described such logic in the "Possible Solution" section

> UX Turbo package should be the one setting the appropriate headers

I also thought about that, but the disadvantage of such an approach would likely be a (barely noticeable) drop in performance, since the WDT is only used in development.

@smnandre So, does that mean you tried to reproduce my issue for real or do I have to provide a bug reproducer?

**Comment by stof** [comment]
@DavidPetrasek TurboBundle could totally register a listener only in debug mode if you don't want this barely noticeable impact at all.

> I am not using morphing

then maybe your blockage is only because of the unsupported CSP

> I have described such logic in the "Possible Solution" section

Your solution relies on storing a CSP nonce in the session and reusing it for next requests, which is **not** what NelmioSecurityBundle does (and would also have a performance impact due to using the session on all pages). This is not a complete solution.

> I'm not sure to be honest, as a standard Symfony + UX Turbo install works without any problem (web debug toolbar included)
> 
> AFAIK, it's only happening with the aditional rules configured in the nelmio security bundle ...

If you don't use a CSP with nonces in it, you will indeed not have issues with Turbo being incompatible with them by default.
The WDT does not force injecting nonces (which would break projects that don't expect a CSP by creating one blocking them). It injects a nonce only in projects that already use CSP (so that the toolbar is not blocked by the CSP).

**Comment by smnandre** [comment]
>  So, does that mean you tried to reproduce my issue for real or do I have to provide a bug reproducer?

No need here, you explained clearly what's happening 👍 


@DavidPetrasek  When i said "i'm not sure",  it was to answer the last sentence from @stof  

>  I think the UX Turbo package should be the one setting [...] 

Because, i may be wrong here but i feel the problem does only occur with the WDT  and the NelmioBundle. 

As the WDT is never available in prod environment, i'm not sure the Ux Turbo Bundle should bear the responsibility to implement such specific code for a dev tool... 
... because someone will need to maintain / document this feature for at least 2 years after.

This is what made me a bit hesitant. 

But i'd be happy to help if I can here  :)

**Comment by stof** [comment]
@smnandre There is 2 different things regarding Turbo and CSP:

- making Turbo compatible with CSP nonces. This is *also* needed in prod, not just in dev, if your policy relies CSP nonces (for instance through the `csp_nonce` function of NelmioSecurityBundle but it could be another source for the nonces)
- setting some `X-SymfonyProfiler-*-Nonce` headers to make the WDT use those nonce values instead of generating its own (needed only in dev). The *values* to use when setting this should come from the logic used for the Turbo integration in the previous point (the code snippet of @DavidPetrasek hardcodes a string literal, but this is totally bad for a nonce as what makes it secure is that it cannot be guessed)

**Comment by smnandre** [comment]
> making Turbo compatible with CSP nonces.

UX Turbo or Turbo itself ? (geniuine question)

**Comment by stof** [comment]
Turbo itself is incompatible with nonces by default, because all suggested solutions to support CSP nonces require a cooperation from the backend server (to make it reuse a previous nonce).
UX Turbo might be the place doing this integration as it can provide both sides of the logic (maybe NelmioSecurityBundle is a better place for adding at least part of the backend logic depending on which bundle should be aware of the other though)

**Comment by DavidPetrasek** [comment]
> (the code snippet of @DavidPetrasek hardcodes a string literal, but this is totally bad for a nonce as what makes it secure is that it cannot be guessed)

I used the value 'fffff' for testing purposes only. The final implementation will, of course, generate a proper nonce.

**Comment by smnandre** [comment]
This is not my strongest domain of expertise..  @DavidPetrasek  would like to work with me on this, and try to make the UX Turbo Bundle better handle this ?

**Comment by DavidPetrasek** [comment]
> This is not my strongest domain of expertise.. @DavidPetrasek would like to work with me on this, and try to make the UX Turbo Bundle better handle this ?

Yes, here is my solution, which doesn't require using any session and also works with morphing.
UX Turbo has to execute this function/method after the page has loaded:
```
symfonyWebDebugToolbar_reuseNonce()
{
    let wdt = document.querySelector('.sf-toolbar');        if (!wdt) {return;}

    let wdtStyle = wdt.nextElementSibling;   
    let wdtScript = wdtStyle.nextElementSibling;
    
    if (!wdtStyle.nonce || !wdtScript.nonce) {return;}

    document.addEventListener('turbo:before-fetch-request', (event) => 
    {           
        event.detail.fetchOptions.headers['X-SymfonyProfiler-Style-Nonce'] = wdtStyle.nonce;
        event.detail.fetchOptions.headers['X-SymfonyProfiler-Script-Nonce'] = wdtScript.nonce;
    });
}
```

Turbo should ideally inject this JS code only when in development environment.
The `before-visit` event would be ideal, but it can't be used because it doesn't allow access to the fetchOptions property. After testing this, I also found that **there is no need** to trigger an update of the WDT by setting the Symfony-Debug-Toolbar-Replace header. **However, the previously mentioned JS errors still occur.**

**Comment by smnandre** [comment]
Ok i'm almost there... made a small repro and added your code.

Once this done,i think i'll expand it to see what can be done for other UX packages (LiveComponent will be something..)

###  Last error  ❌

The last error (that does not seem to be blocking anything) is apparently caused by the _wdt that does not get the nonce set from the page request, and evaluates some script.

<img width="1521" alt="csp-turbo" src="https://github.com/user-attachments/assets/057aab0a-ece2-4b1c-b32c-4026ba57fa7f">

### Working ✅ 

The code you provided does indeed work very well. I only had to use the "hash trick" to avoid the style violation cause by the Turbo script injecting the progress styles... it feels a bit unstable / hacky to be done in a bundle .. 😄  

Let's try to solve this last error  and see where to add your code. As you said, it should only be set in dev (i'm tempted to say: only in _debug_ mode and _if_ the wdt is enabled.

I'll poke you tomorrow if i have time to finish this little experiment :)

**Comment by smnandre** [comment]
Not sure here, but it feels _like_ the updateResponseHeaders() is called after the wdt is rendered (in AJAX) and does not set the same nonces used to render the wdt. If so, we're probably missing a second event listener in the script.

**Comment by smnandre** [comment]
I confirm it's caused by to the XHR call (that does not take options for now, and is a bit deep in the profiler code (in toolbar_js.html.twig) but i guess a nonce could be injected.. or fetched before request is called)

**Comment by DavidPetrasek** [comment]
> I only had to use the "hash trick" to avoid the style violation cause by the Turbo script injecting the progress styles... it feels a bit unstable / hacky to be done in a bundle
Do you mean this? `<meta name="csp-nonce" content="..whatever_nonce...">`

> and if the wdt is enabled
Yes, that would be great.

Since the mentioned errors are not triggered on each page refresh (on some refreshes there are no errors) I checked the toolbar_js.html.twig and found it uses `onreadystatechange` event which I believe is the culprit here, because this event can fire when document has not finished loading and therefore I got `ajaxToolbarPanel is null`

**Comment by DavidPetrasek** [comment]
> and if the wdt is enabled

Yes, that would be great.

> If so, we're probably missing a second event listener in the script.

I checked the requests being made, and I confirm that after visiting a link, **only two requests are made every time**, the second being the WDT. However, the mentioned errors **do not occur on every Turbo Drive visit.**

> I confirm it's caused by to the XHR call

So you think the WDT is making AJAX calls on page load?
That's not what I see — when the error occurs, **only two requests** are made.

**Comment by smnandre** [comment]
Yes, and for me the one blocked (that display an error) is the XHR call and it is not handled by Stimulus/Turbo, but here:

https://github.com/symfony/symfony/blob/f91514df907764290a70812fd7a0600be6e26c38/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig#L57

**Comment by smnandre** [comment]
> However, the mentioned errors do not occur on every Turbo Drive visit.

I _**think**_ because Turbo does not refresh the wdt code / reexecute the script every time.

**Comment by DavidPetrasek** [comment]
I do not really understand why WDT is being added twice and why it also uses the `onreadystatechange` event. 
Since this event **can fire multiple times** on page load I think it could be the culprit.

**Comment by DavidPetrasek** [comment]
I completely disabled the CSP and after a visiting few links the error `ajaxToolbarPanel is null` appeared again. This happened multiple times.

**Comment by smnandre** [comment]
That does not surprise me... i'm now 99% certain it's due to the "retry" strategy of XHR. Comes a moment when a late answer land later than a request... and this start the mess with the nonces.

**Comment by SimonChabrier** [comment]
Hello, this is indeed a big, very annoying problem with UxTurbo and the debugBar if we make strong CSP.

I will share with you my solution which is radical but works very well in symfony 7.2. We consider of course that if we have the debug bar it is that we are in development and that we never activate the debug bar in production.

Have a look in vendor, there is `'Symfony\Bundle\WebProfilerBundle\Csp\ContentSecurityPolicyHandler'` and the class have a `public function disableCsp()`

So the solution is : 

## 1 Make a Service 

```php
<?php

namespace App\Service\System;

use Symfony\Component\DependencyInjection\Attribute\When;
use Symfony\Bundle\WebProfilerBundle\Csp\ContentSecurityPolicyHandler;

#[When(env: 'dev')]
class DesableDebubBarCsp
{

    public function __construct(
        private ContentSecurityPolicyHandler $contentSecurityPolicyHandler,
    ) {}

    /**
     * Hoverhide ContentSecurityPolicyHandler from bundle symfony/web-profiler-bundle
     * @return void
     */
    public function disableCsp(): void
    {
        $this->contentSecurityPolicyHandler->disableCsp();
    }
}
```

## 2 Add an alias in services.yaml 

```Symfony\Bundle\WebProfilerBundle\Csp\ContentSecurityPolicyHandler: '@web_profiler.csp.handler'```

## 3 Make the Subscriber to inject the debug bar 

```php
<?php

namespace App\EventSubscriber\System;

use App\Service\System\DesableDebubBarCsp;
use Symfony\Component\HttpKernel\KernelEvents;
use Symfony\Component\HttpKernel\KernelInterface;
use Symfony\Component\HttpKernel\Event\ResponseEvent;
use Symfony\Component\DependencyInjection\Attribute\When;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;

#[When(env: 'dev')]
class AjaxDebugBarSubscriber implements EventSubscriberInterface
{

    public function __construct(
        private KernelInterface $kernel,
        private DesableDebubBarCsp $csp
    ) {}

    public function onKernelResponse(ResponseEvent $event): void
    {
        // this remove the nonce generation
        $this->csp->disableCsp();

        if (!$this->kernel->isDebug()) {
            return;
        }

        $request = $event->getRequest();
        if (!$request->isXmlHttpRequest()) {
            return;
        }

        $response = $event->getResponse();

        $response->headers->set('Symfony-Debug-Toolbar-Replace', '1');
    }

    public static function getSubscribedEvents(): array
    {
        return [
            KernelEvents::RESPONSE => 'onKernelResponse',
        ];
    }
}
```

Finally, as the debug bar is injected after Nelmio Security has evaluated the state of the DOM (I thinks this), there are no more errors in the console after Ajax responses and we can continue to NOT ADD 'unsafe-inline' in the Nelmio Security configuration and continue to evaluate other possible problems related to other things without being bothered by the debug bar.

I tried differently before, someone had shared a small Javascript solution but it does not work systematically according to the rendering times and the problems persist. Here this solution is radical and for my part has completely solved the problem. I hope it will help you.

Do not forget that this is a voluntarily opened flaw and that the debugbar must never be activated in production for any reason.

That said, it is a real problem and it would be great if the symfony teams soon proposed a real solution that allows the injection of the debug bar in ajax responses to communicate with Turbo so that it is automated and does not force us to do radial processing like that.

You will tell me if it worked for you if someone decides to adopt this approach.

**Comment by DavidPetrasek** [comment]
@SimonChabrier  Few months ago I "solved it" by just not using CSP in dev environment. I appreciate your solution, but I will not try that since its not a real solution. As you can see in the previous comments the problem most likely lies inside this file: [symfony/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig](https://github.com/symfony/symfony/blob/f91514df907764290a70812fd7a0600be6e26c38/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar_js.html.twig#L57)
... so whoever wrote this file should take a look a it. I invite @fabpot to take a look at it.

@SimonChabrier  Please avoid duplicate comments, so its better if you delete your duplicate comment from here https://github.com/symfony/ux/issues/2134 and just mention that issue in your comment inside this issue.

**Comment by SimonChabrier** [comment]
Hello. You say "Few months ago I "solved it" by just not using CSP in dev environment" I did not understand where is your solution. Could you explain me please ? (I delete my contribution in the other post.)

**Comment by SimonChabrier** [comment]
@DavidPetrasek  Hello. You say "Few months ago I "solved it" by just not using CSP in dev environment" I did not understand where is your solution. Could you explain me please ? (I delete my contribution in the other post.)

**Comment by DavidPetrasek** [comment]
> [@DavidPetrasek](https://github.com/DavidPetrasek) Hello. You say "Few months ago I "solved it" by just not using CSP in dev environment" I did not understand where is your solution. Could you explain me please ? (I delete my contribution in the other post.)

The "solution" for me was to stop using CSP in dev environment since CSP is not needed in dev.

**Comment by SimonChabrier** [comment]
> The "solution" for me was to stop using CSP in dev environment since CSP is not needed in dev.


Ok I thought you had proposed a solution and I had not seen it. In any case if you use symfony 7.2 and UxTurbo you can implement my proposal in your code and continue to enjoy your CSP in development environment. It is tested and functional. The proposal simply uses a public method of the php class and only works on the debug bar and in dev environment as we see it until the new attribute to target the development environment only. I am sure and hope it will help people while waiting for the authors to propose a native system to solve this problem because for me these places are also used for that.

**Comment by DavidPetrasek** [comment]
I myself already created a fully working solution here (https://github.com/symfony/symfony/issues/58223#issuecomment-2351509377) which is much simpler than yours. Only minor console errors needs to be fixed (https://github.com/symfony/symfony/issues/58223#issuecomment-2360043257)

**Comment by SimonChabrier** [comment]
> I myself already created a fully working solution here (https://github.com/symfony/symfony/issues/58223#issuecomment-2351509377) which is much simpler than yours. Only minor console errors needs to be fixed (https://github.com/symfony/symfony/issues/58223#issuecomment-2360043257)

yes I saw it and tested it but it does not work 100%. Because when you use Turbo pre-fetch on links and you hover over several links without direct action, then when you return to a link to finally do the action the ajax request is not redone and we are no longer synchronized so a CSP error is raised. But your solution is the first thing I tested and it also helped me to better understand the situation. ;)

**Comment by DavidPetrasek** [comment]
> yes I saw it and tested it but it does not work 100%.

Yes there is a CSP violation as mentioned at the beginning (`Content-Security-Policy: The page’s settings blocked an inline script (script-src-elem) ...`) and as mentioned here (https://github.com/symfony/symfony/issues/58223#issuecomment-2351509377) and it still do occur, but but it happens even when pre-fetch is disabled on specific link.

Anyway, was trying to solve it.
I have updated the code to:
```
document.addEventListener('turbo:before-fetch-request', (event) => 
{
    var wdt = document.querySelector('.sf-toolbar');
    if (wdt)
    {
        let wdtStyle = wdt.nextElementSibling;
        let wdtScript = wdtStyle.nextElementSibling;

        if (wdtStyle.nonce) {event.detail.fetchOptions.headers['X-SymfonyProfiler-Style-Nonce'] = wdtStyle.nonce;}
        if (wdtScript.nonce) {event.detail.fetchOptions.headers['X-SymfonyProfiler-Script-Nonce'] = wdtScript.nonce;}
    }
});
```

**Comment by SimonChabrier** [comment]
You should try my solution it is rendered server side and uses a native method of the class so it is more robust than working client side. It is not complex it just takes the script from the documentation which explains how to handle the refresh of the debug bar after Ajax requests. We just add the call to the bundle method in a custom Service to disable the CSP nothing more.

The doc script is here : https://symfony.com/doc/current/profiler.html#updating-the-web-debug-toolbar-after-ajax-requests

I just added the call to the bundle method so that it is done correctly and supports a bundle update without changing the expected result.

**Comment by DavidPetrasek** [comment]
@SimonChabrier I have investigated and tried your solution. It is not a solution at all. What is does is, **it disables the CSP completely**, so that's why you don't see any violations. See the function and what it does:
```
/**
     * Disables Content-Security-Policy.
     *
     * All related headers will be removed.
     */
    public function disableCsp(): void
    {
        $this->cspDisabled = true;
    }
```

This is part of the same class I've just made edits to (see my previous answer).
@stof @smnandre @xabbuh And by the way, this seems like a mistake and should be removed from `ContentSecurityPolicyHandler` class together with `removeCspHeaders` method or moved somewhere else, because again this disables CSP not only for the profiler bundle, but for the whole app. The profiler bundle should not be able to disable the CSP completely. Prove me wrong.

**Comment by SimonChabrier** [comment]
You're right, you saw it right. I hadn't noticed that the scope was global. So to solve the problem I basically used the nonce generated by Nelmio directly. Since it generates it itself for each response, it recognizes it and authorizes it. I then tested commenting on scripts authorized in Nelmio CSPs and they do return an error, but the Nelmio nonce is injected into the value of the Debug Bar nonce, so no more worries. With your comment and a fix from me, I think we've found the right solution.

```php
<?php

namespace App\Service\System;

use Symfony\Component\DependencyInjection\Attribute\When;
use Nelmio\SecurityBundle\ContentSecurityPolicy\NonceGenerator;

#[When(env: 'dev')]
class NelmioNonceService
{

    public function __construct(
        private NonceGenerator $nonceGenerator,
    ) {}

    /**
     * Get the CSP nonce from the Nelmio SecurityBundle
     * @return string
     */
    public function getNelmioNonce(): string
    {
        return $this->nonceGenerator->generate('script'); // Devrait être le même que `csp_nonce('script')`
    }
}
````

services.yaml :     Nelmio\SecurityBundle\ContentSecurityPolicy\NonceGenerator: '@nelmio_security.nonce_generator'

(to show all Nelmio services id : bin/console debug:container | grep nelmio)

And finally : 

```php
<?php

namespace App\EventSubscriber\System;

use App\Service\System\DesableDebubBarCsp;
use App\Service\System\NelmioNonceService;
use Symfony\Component\HttpKernel\KernelEvents;
use Symfony\Component\HttpKernel\KernelInterface;
use Symfony\Component\HttpKernel\Event\ResponseEvent;
use Symfony\Component\DependencyInjection\Attribute\When;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;

// https://symfony.com/doc/current/profiler.html#updating-the-web-debug-toolbar-after-ajax-requests
// Classe qui permet de mettre à jour la barre de debug symfony après une requête ajax en ajoutant un header à la réponse
#[When(env: 'dev')]
class AjaxDebugBarSubscriber implements EventSubscriberInterface
{

    public function __construct(
        private KernelInterface $kernel,
        private DesableDebubBarCsp $csp,
        private NelmioNonceService $nonceService
    ) {}

    public function onKernelResponse(ResponseEvent $event): void
    {

        if (!$this->kernel->isDebug()) {
            return;
        }

        $request = $event->getRequest();
        if (!$request->isXmlHttpRequest()) {
            return;
        }

        $response = $event->getResponse();

        // get the CSP nonce from the Nelmio SecurityBundle
        $nelmioScriptNonce = $this->nonceService->getNelmioNonce();

        // Replace the Symfony Debug Toolbar nonce with the Nelmio SecurityBundle nonce
        if ($response->headers->has('x-symfonyprofiler-script-nonce')) {
            $response->headers->set('x-symfonyprofiler-script-nonce', $nelmioScriptNonce);
        }

        if ($response->headers->has('x-symfonyprofiler-style-nonce')) {
            $response->headers->set('x-symfonyprofiler-style-nonce', $nelmioScriptNonce);
        }

        $response->headers->set('Symfony-Debug-Toolbar-Replace', '1');
    }

    public static function getSubscribedEvents(): array
    {
        return [
            KernelEvents::RESPONSE => 'onKernelResponse',
        ];
    }
}
````

**Comment by SimonChabrier** [comment]
You are absolutely right I had not seen that the scope was general. I have not found any way to achieve a result. I tried to decorate the native service but it does not work either. It is a headache.

**Comment by DavidPetrasek** [comment]
**I have solved the CSP errors by resolving other issue I had. It is all explained here:** https://github.com/hotwired/turbo/issues/1366#issuecomment-2660852641

However, the already mentioned error `ajaxToolbarPanel is null` still occasionally occurs.

**Comment by SimonChabrier** [comment]
I spent a long time on it, for my part I stop looking. Everything I could try did not give a stable result. I will stick to what I did even if it globally deactivates the CSP. It only works in a development environment, I will be satisfied with that even if it is very frustrating...

**Comment by DavidPetrasek** [comment]
Created a new PR to solve this issue: https://github.com/symfony/ux/pull/2633
